### PR TITLE
Add The Ten Years Bugs

### PR DESCRIPTION
**Pull request by alexpusch**

### CODE REVIEW

File: the-then-years-bug/the-ten-years-bug.md [hunk 1]
```diff
@@ -0,0 +1,86 @@
+The Ten Years Bug
+=========================
+
+Developers often make mistakes. Bugs goes unnoticed, uncaught by unitests, overlooked by QA and reach production. Sometimes it's just a minor issue, sometimes it's major bug that costs you lost data and users confidence. Whatever it is, the issue gets fixed. You patch the bug, mend corrupt data and everything get back to it's place.
```
Comment (itayadler) [inline the-then-years-bug/the-ten-years-bug.md]: "Bugs goes unnoticed" => "Bugs go unnoticed"

Comment (itayadler) [inline the-then-years-bug/the-ten-years-bug.md]: "get back to it's place" => "gets back to its place"

File: the-then-years-bug/the-ten-years-bug.md [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+Surly there isn't a bug that is unfixable right?
```
Comment (itayadler) [inline the-then-years-bug/the-ten-years-bug.md]: Surly => Surely

File: the-then-years-bug/the-ten-years-bug.md [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+Our snippet
+------------
+As a prologue to this story I'll tell you about our snippet. To use [Jaco](https://getjaco.com), just sign up, copy a small snippet to your site, and your users sessions starts to get recorded. The main goal of the snippet is this:
```
Comment (itayadler) [inline the-then-years-bug/the-ten-years-bug.md]: starts => start

File: the-then-years-bug/the-ten-years-bug.md [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+}
+```
+
+There is a horrible reality hidden in this piece of code. The Cache-Control max-age is set to 315360000. 315360000 seconds, 5256000 minutes, 87600 hours, 3650 days. 10 years.
+
+What is Cache-Control?
+------------
+[To quote Googles performance guide](https://developers.google.com/web/fundamentals/performance/optimizing-content-efficiency/http-caching#cache-control) "The best request is a request that does not need to communicate with the server". Cache-Control the modern method to customize the browsers caching policy.
+
+This policy aims to help the browser to answer questions like
```
Comment (itayadler) [inline the-then-years-bug/the-ten-years-bug.md]: -to

File: the-then-years-bug/the-ten-years-bug.md [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+
+In other words we just told the all of our end users browsers to cache our recorder for 10 years without downloading new versions, or doing any kind of communication with S3 to check for a change in the caching policy. All the minor bugs we fixed in the past still linger in our end users caches, and worse, we don't have anything to do about it. Fuck.
+
+What we did
+-----------
+Under different circumstances this won't be a big issue. Just change the filename and be done with it. But for our case, changing the filename means changing the snippet, and changing the snippet means contacting every client, and asking nicely to update their code.
+
+Well it couldn't be helped. We've created recorder_v2.js, published an email with an explanation and instruction how to change the snippet. As an auxiliary fix we've configured S3 to permanent redirect all request to recorder.js to recorder_v2.js
+
+These actions decreased the amount of old recorder.js file being used, but not eliminated it completely. Some clients are less cooperative with request such as ours and did not updated their snippet. Theoretically some bugs will hunt us for another 10 years.
```
Comment (itayadler) [inline the-then-years-bug/the-ten-years-bug.md]: Please remove this paragraph

File: the-then-years-bug/the-ten-years-bug.md [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+
+In conclusion
+------------
```
Comment (itayadler) [inline the-then-years-bug/the-ten-years-bug.md:104]: cacheing => caching

Comment (itayadler) [inline the-then-years-bug/the-ten-years-bug.md:104]: check https://jakearchibald.com/2016/caching-best-practices/

File: the-then-years-bug/the-ten-years-bug.md (hunk 1 above)
Comment (alexpusch) [reply the-then-years-bug/the-ten-years-bug.md]: done

File: the-then-years-bug/the-ten-years-bug.md (hunk 2 above)
Comment (alexpusch) [reply the-then-years-bug/the-ten-years-bug.md]: done

File: the-then-years-bug/the-ten-years-bug.md (hunk 3 above)
Comment (alexpusch) [reply the-then-years-bug/the-ten-years-bug.md]: done

File: the-then-years-bug/the-ten-years-bug.md (hunk 5 above)
Comment (alexpusch) [reply the-then-years-bug/the-ten-years-bug.md]: ok

File: the-then-years-bug/the-ten-years-bug.md [hunk 7]
```diff
@@ -1,13 +1,13 @@
 The Ten Years Bug
 =========================
 
-Developers often make mistakes. Bugs goes unnoticed, uncaught by unitests, overlooked by QA and reach production. Sometimes it's just a minor issue, sometimes it's major bug that costs you lost data and users confidence. Whatever it is, the issue gets fixed. You patch the bug, mend corrupt data and everything get back to it's place.
+Developers often make mistakes. Bugs go unnoticed, uncaught by unit tests, overlooked by QA and reach production. Sometimes it's just a minor issue, sometimes it's major bug that costs you lost data and users' confidence. Whatever it is, the issue gets fixed. You patch the bug, mend corrupt data and everything gets back to its place.
 
 Surly there isn't a bug that is unfixable right?
```
Comment (itayadler) [inline the-then-years-bug/the-ten-years-bug.md]: surly => surely

**Comment by alexpusch** [comment]
- Add what we should have done, max-age: [small number] or no-cache. no-store is no good

**Pull request merged**